File: 프로그래머스/PCCP/석유 시추.cpp
// https://school.programmers.co.kr/learn/courses/30/lessons/250136

#include <string>
#include <vector>
#include <queue>
#include <set>

using namespace std;

int n = 500, m = 500;
const int dy[4] = {-1, 1, 0, 0};
const int dx[4] = {0, 0, -1, 1};

bool visited[500][500];
vector<vector<int>> board;
vector<pair<int, set<int>>> areas; // 넓이, col 좌표들

int max(int a, int b) { return a > b ? a : b; }

void BFS(int y, int x) {
    int ret = 1;
    queue<pair<int,int>> que;
    set<int> areaCol;
    
    que.push({y,x});
    areaCol.insert(x);
    visited[y][x] = true;
    
    while (!que.empty()) {
        int cy = que.front().first;
        int cx = que.front().second;
        que.pop();
        
        for (int dir = 0; dir < 4; ++dir) {
            int ny = cy + dy[dir];
            int nx = cx + dx[dir];
            
            if (ny < 0 || nx < 0 || ny >= n || nx >= m) {
                continue;
            }
            
            if (board[ny][nx] == 0 || visited[ny][nx]) { continue; }
            
            que.push({ny, nx});
            visited[ny][nx] = true;
            ret++;
            areaCol.insert(nx);
        }
    }

    areas.push_back({ret, areaCol});
    return;
}

int solution(vector<vector<int>> land) {
    int answer = 0;
    n = land.size(), m = land[0].size();
    board = land;
    
    for (int i = 0; i < n; ++i) {
        for (int j = 0; j < m; ++j) {
            if (land[i][j] == 0 || visited[i][j]) {
                continue;
            }
            BFS(i, j);
        }
    }
    
    int colAmount[501] = {0};
    for (int i; i < areas.size(); ++i) {
        for (auto it: areas[i].second) {
            colAmount[it] += areas[i].first;
            answer = max(answer, colAmount[it]);
        }
    }
    
    return answer;
}

File: 프로그래머스/PCCP/석유 시추.py
# https://school.programmers.co.kr/learn/courses/30/lessons/250136
from collections import deque

visited = [[0] * 500 for i in range(500)]
oilGrup = []

row, col = 500, 500

def BFS(land, x, y):
    global oilGrup, visited
    
    ret = 1
    colSet = set()
    que = deque([])
    
    dir = [[1, 0], [-1, 0], [0, 1], [0, -1]]
    que.append((x, y))
    colSet.add(y)
    visited[x][y] = True
    
    while que:
        cx, cy = que.popleft()
        
        for d in dir:
            nx = cx + d[0]
            ny = cy + d[1]
            
            if nx < 0 or ny < 0 or nx >= row or ny >= col:
                continue
            
            if visited[nx][ny] or land[nx][ny] == 0: 
                continue
            
            ret += 1
            visited[nx][ny] = True
            colSet.add(ny)
            que.append((nx, ny))
    
    
    oilGrup.append((ret, colSet))
    
    return

def init(land):
    global row, col
    
    row, col = len(land), len(land[0])
    
    return

def makeOilGruop(land):
    for i in range(row):
        for j in range(col):
            if land[i][j] == 1 and not visited[i][j]:
                BFS(land, i, j)
                
    return

def solution(land):
    answer = 0
    
    init(land)
    makeOilGruop(land)
    
    colAmounts = [0] * col
    #print(oilGrup)
    
    for oilSize, colSet in oilGrup:
        for c in colSet:
            colAmounts[c] += oilSize
            answer = max(colAmounts[c], answer)
        
    #print(colAmounts)
    return answer
